jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourceTransformer.java

```diff
@@ -122,25 +145,68 @@ public Content.Chunk read()
                         transformedChunk = Content.Chunk.asChunk(transformedChunk.getByteBuffer(), false, transformedChunk);
                 }
 
-                boolean terminated = rawLast && transformedLast;
-                boolean terminalFailure = transformedFailure && transformedLast;
-
-                Content.Chunk result = transformedChunk;
-                transformedChunk = Content.Chunk.next(result);
+                if (transformedFailure && transformedLast)
+                    return afterRead(State.Type.FAILED, transformedChunk);
 
-                if (terminated || terminalFailure)
-                {
-                    finished = true;
-                    release();
-                }
+                if (rawLast && transformedLast)
+                    return afterRead(State.Type.EOF, transformedChunk);
 
-                return result;
+                return afterRead(State.Type.IDLE, transformedChunk);
             }
 
             needsRawRead = rawChunk == null;
         }
     }
 
+    private Content.Chunk afterRead(State.Type type, Content.Chunk chunk)
+    {
+        while (true)
+        {
+            State current = state.get();
+            switch (current.type)
+            {
+                case IDLE, EOF, FAILED -> throw new IllegalStateException();
+                case READING ->
+                {
+                    switch (type)
+                    {
+                        case IDLE ->
```

Comment:
It's confusing when reading to have `State` constants named like `State.Type` values. Maybe use a prefix for the `State` constants to help discriminate them at sight?